File: CPP/ByTier/H/1873_H_Mad_City.cpp
// Problem: CF 1873 H - Mad City
// https://codeforces.com/contest/1873/problem/H

/*
Purpose: 
This code determines whether Valeriu can escape Marcel forever in a game played on a graph representing a city with buildings and roads. 
Marcel and Valeriu start at different buildings and move optimally. Valeriu knows Marcel's move before deciding his own.

Algorithm:
1. Build an adjacency list representation of the graph.
2. Find a cycle in the graph using DFS (the 'en' variable marks the start of the cycle).
3. Perform BFS from the beginning of the cycle to compute distances from that node.
4. Compare the distance from Marcel's starting position to the cycle with that of Valeriu's starting position.
   If Marcel is farther from the cycle than Valeriu, Valeriu can escape, otherwise, Marcel can catch him.

Time Complexity: O(n) per test case due to linear traversal of the graph for DFS and BFS.
Space Complexity: O(n) for storing the graph and auxiliary arrays.

*/
#include <stdio.h>
#include <algorithm>
#include <queue>

using namespace std;
const int N = 400010;
int n, s, t, idx, en, h[N], e[N], d[N], ne[N], st[N];

// Add edge to adjacency list
void add(int a, int b) { 
    e[idx] = b, ne[idx] = h[a], h[a] = idx++; 
}

// DFS to detect a cycle in the graph
void dfs(int u, int p) {
  st[u] = 1;
  for (int i = h[u]; ~i; i = ne[i]) {
    int j = e[i];
    if (j == p)
      continue;
    if (en)
      return;
    if (st[j]) {
      en = j;
      return;
    }
    if (!st[j])
      dfs(j, u);
  }
  return;
}

// BFS to calculate distances from a given node
void bfs(int u) {
  queue<int> q;
  q.push(u);
  while (q.size()) {
    u = q.front();
    q.pop();
    st[u] = 1;
    for (int i = h[u]; ~i; i = ne[i]) {
      int j = e[i];
      if (st[j])
        continue;
      st[j] = 1;
      d[j] = d[u] + 1;
      q.push(j);
    }
  }
  return;
}

// Solve function for each test case
void solve() {
  en = 0;
  scanf("%d%d%d", &n, &s, &t);
  for (int i = 1; i <= n; i++)
    h[i] = -1, st[i] = 0;
  for (int i = 0; i < n; i++) {
    int a, b;
    scanf("%d%d", &a, &b);
    add(a, b);
    add(b, a);
  }
  for (int i = 1; i <= n; i++)
    st[i] = d[i] = 0;
  dfs(t, 0);  // Find the cycle starting from Valeriu's position
  for (int i = 1; i <= n; i++)
    st[i] = 0;
  bfs(en);  // BFS from the start of the cycle to find all distances
  if (d[s] > d[t])
    printf("YES\n");
  else
    printf("NO\n");
}

// Main function to read input and test cases
int main() {
  int T;
  scanf("%d", &T);
  while (T--) {
    solve();
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1873/1873_H_Mad_City.py
# Problem: CF 1873 H - Mad City
# https://codeforces.com/contest/1873/problem/H

"""
Algorithm: Determine if Valeriu can escape Marcel forever in a graph-based game.
Techniques: 
- Graph processing to identify cycle nodes using BFS (removing leaves iteratively)
- BFS from both starting positions to find distances to cycle nodes
- Compare distances to assess escape possibility

Time Complexity: O(n + m) where n is number of nodes and m is number of edges
Space Complexity: O(n + m) for adjacency list representation and tracking arrays
"""

import sys
from collections import deque

input = sys.stdin.readline


def can_escape(n, a, b, edges):
    # Build adjacency list and degree count
    g = [[] for _ in range(n + 1)]
    deg = [0] * (n + 1)
    for u, v in edges:
        g[u].append(v)
        g[v].append(u)
        deg[u] += 1
        deg[v] += 1
    
    # Identify nodes not on cycles by removing leaves iteratively
    q = deque()
    on_cycle = [True] * (n + 1)  # Initially all nodes are considered on cycle
    for i in range(1, n + 1):
        if deg[i] == 1:
            q.append(i)
            on_cycle[i] = False  # Leaf nodes are not on cycle
    
    # Remove leaves iteratively until only cycles remain
    while q:
        leaf = q.popleft()
        for nei in g[leaf]:
            if on_cycle[nei]:
                deg[nei] -= 1
                if deg[nei] == 1:
                    on_cycle[nei] = False
                    q.append(nei)
    
    # BFS from Valeriu's position to find nearest cycle node
    dist_b = [-1] * (n + 1)
    qb = deque([b])
    dist_b[b] = 0
    meet = -1  # Target node to meet at (first cycle node encountered)
    while qb:
        u = qb.popleft()
        if on_cycle[u]:
            meet = u
            break
        for v in g[u]:
            if dist_b[v] == -1:
                dist_b[v] = dist_b[u] + 1
                qb.append(v)
    
    # BFS from Marcel's position to find distance to the same target node, if possible
    dist_a = [-1] * (n + 1)
    qa = deque([a])
    dist_a[a] = 0
    while qa and dist_a[meet] == -1:  # Continue while not reached meet node
        u = qa.popleft()
        for v in g[u]:
            if dist_a[v] == -1:
                dist_a[v] = dist_a[u] + 1
                qa.append(v)
    
    # If Valeriu can reach the meeting point before or at same time as Marcel, he escapes
    return dist_b[meet] < dist_a[meet]


def main():
    t = int(input())
    out = []
    for _ in range(t):
        n, a, b = map(int, input().split())
        edges = [tuple(map(int, input().split())) for _ in range(n)]
        out.append("YES" if can_escape(n, a, b, edges) else "NO")
    print("\n".join(out))


if __name__ == "__main__":
    main()


# https://github.com/VaHiX/CodeForces/